### PR TITLE
UCS/LOG: Store an allocated filename in log filter key hash

### PR DESCRIPTION
**Pull request by tvegas1**
## What
Make the global log filter key hash store allocated filenames.

## Why ?
At lookup time, comparing against a previously stored `__FILE__` leads to `SEGFAULT` if the corresponding library has already been `dlclose()`'d.

## How ?
Use variable size chunk allocation.

Gtest are missing, if approach is ok, we can move this to `ucs/datastruct`, and add corresponding tests.

### Crash
configure: `--enable-logging=trace_poll`
run: `mpirun -x UCX_LOG_LEVEL=trace -x UCX_PROTO_ENABLE=y -mca pml ucx -np 2 osu_bw -x 400 -m 4096:4096`

```
[jazz29:39743] *** Process received signal ***
[jazz29:39743] Signal: Segmentation fault (11)
[jazz29:39743] Signal code:  (-6)
[jazz29:39743] Failing at address: 0x182cd00009b3f
[jazz29:39743] [ 0] /lib64/libpthread.so.0(+0xf5d0)[0x7f13f847e5d0]
[jazz29:39743] [ 1] /lib64/libc.so.6(+0x13e016)[0x7f13f81e0016]
[jazz29:39743] [ 2] /labhome/tvegas/share/srcs/hpcx/modules/ucx/rfs/lib/libucs.so.0(ucs_log_default_handler+0xe9)[0x7f13e91331b9]
[jazz29:39743] [ 3] /labhome/tvegas/share/srcs/hpcx/modules/ucx/rfs/lib/libucs.so.0(ucs_log_dispatch+0xdf)[0x7f13e9133d0f]
[jazz29:39743] [ 4] /labhome/tvegas/share/srcs/hpcx/modules/ucx/rfs/lib/libucs.so.0(+0x4ba4b)[0x7f13e911ca4b]
[jazz29:39743] [ 5] /labhome/tvegas/share/srcs/hpcx/modules/ucx/rfs/lib/libucs.so.0(ucs_async_global_cleanup+0x33)[0x7f13e911add3]
[jazz29:39743] [ 6] /labhome/tvegas/share/srcs/hpcx/modules/ucx/rfs/lib/libucs.so.0(+0x44039)[0x7f13e9115039]
[jazz29:39743] [ 7] /lib64/ld-linux-x86-64.so.2(+0xffca)[0x7f13f8ccbfca]
[jazz29:39743] [ 8] /lib64/libc.so.6(+0x39b69)[0x7f13f80dbb69]
[jazz29:39743] [ 9] /lib64/libc.so.6(+0x39bb7)[0x7f13f80dbbb7]
[jazz29:39743] [10] /lib64/libc.so.6(__libc_start_main+0xfc)[0x7f13f80c43dc]
[jazz29:39743] [11] /hpc/local/benchmarks/daily/next/2023-06-19/hpcx-gcc-redhat7/ompi/tests/osu-micro-benchmarks-5.8/osu_bw[0x401df0]
[jazz29:39743] *** End of error message ***
==== backtrace (tid:  39744) ====
 0 0x000000000013e016 __strcmp_sse42()  :0
 1 0x00000000000621b9 kh_str_hash_equal()  /labhome/tvegas/share/srcs/hpcx/modules/ucx/src/ucs/datastruct/khash.h:456
 2 0x000000000004ba4b ucs_async_signal_global_cleanup()  /labhome/tvegas/share/srcs/hpcx/modules/ucx/src/ucs/async/thread.c:433
 3 0x0000000000049dd3 ucs_async_global_cleanup()  /labhome/tvegas/share/srcs/hpcx/modules/ucx/src/ucs/async/async.c:682
 4 0x0000000000044039 ucs_cleanup()  /labhome/tvegas/share/srcs/hpcx/modules/ucx/src/ucs/sys/init.c:130
 5 0x000000000000ffca _dl_fini()  :0
 6 0x0000000000039b69 __run_exit_handlers()  :0
 7 0x0000000000039bb7 __GI_exit()  :0
 8 0x00000000000223dc __libc_start_main()  ???:0
 9 0x0000000000401df0 _start()  ???:0
=================================
```

### CODE REVIEW

File: src/ucs/debug/log.c [hunk 1]
```diff
@@ -296,10 +355,17 @@ ucs_log_default_handler(const char *file, unsigned line, const char *function,
         /* Add source file name to the hash */
         match  = fnmatch(ucs_global_opts.log_component.file_filter, file, 0) !=
                  FNM_NOMATCH;
-        khiter = kh_put(ucs_log_filter, &ucs_log_global_filter, file, &khret);
-        ucs_assert((khret == UCS_KH_PUT_BUCKET_EMPTY) ||
-                   (khret == UCS_KH_PUT_BUCKET_CLEAR));
-        kh_val(&ucs_log_global_filter, khiter) = match;
+
+        filename = ucs_mem_chunk_alloc_str(file);
```
Comment (yosefe) [inline src/ucs/debug/log.c:359]: why not just use ucs_malloc/ucs_strdup?

Comment (tvegas1) [reply src/ucs/debug/log.c:359]: small allocations that don't need individual tracking, shall we use:
- create: `ucs_strdup()`
- destroy: `kh_foreach_key() { ucs_free() }`, then destroy kh itself

Comment (yosefe) [reply src/ucs/debug/log.c:359]: Yes

Comment (tvegas1) [reply src/ucs/debug/log.c]: fixed, retested final patch, kh foreach needs `const char *` buf `ucs_free` needs unconstified ptr.

**Review by brminich — APPROVED** [review]

File: src/ucs/debug/log.c [hunk 2]
```diff
@@ -296,7 +296,14 @@ ucs_log_default_handler(const char *file, unsigned line, const char *function,
         /* Add source file name to the hash */
         match  = fnmatch(ucs_global_opts.log_component.file_filter, file, 0) !=
                  FNM_NOMATCH;
-        khiter = kh_put(ucs_log_filter, &ucs_log_global_filter, file, &khret);
+
+        filename = ucs_strdup(file, "log filter filename");
```
Comment (yosefe) [inline src/ucs/debug/log.c:300]: maybe replace it by ucs_string_set_t?

Comment (tvegas1) [reply src/ucs/debug/log.c:300]: 1- Currently `ucs_log_global_filter` is used to cache if a given file is a match or not. That boolean is accessed with `kh_val()`. But with `ucs_string_set_t` we can only track presence, so we would need to add one set to track caching state (already cached or not), and another set to track matching value (true or false).

2- If suggestion is to add `ucs_string_set_t` to handle memory allocation/tracking only, seems it does not expose string pointers, seems the function could be added though.

3- IMHO seems simpler like it currently is, let me know which option or any other.

Comment (yosefe) [reply src/ucs/debug/log.c:300]: ok

**Review by yosefe — APPROVED** [review]
@tvegas1 can u pls squash?

**Review by yosefe — APPROVED** [review]

**Pull request merged**